mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraBlobsDAO.java

```diff
@@ -0,0 +1,177 @@
+        this.dataChunker = new DataChunker();
+        this.insert = prepareInsert(session);
+        this.select = prepareSelect(session);
+
+        this.insertPart = prepareInsertPart(session);
+        this.selectPart = prepareSelectPart(session);
+    }
+
+    private PreparedStatement prepareSelect(Session session) {
+        return session.prepare(select()
+            .from(Blobs.TABLE_NAME)
+            .where(eq(Blobs.ID, bindMarker(Blobs.ID))));
+    }
+
+    private PreparedStatement prepareSelectPart(Session session) {
+        return session.prepare(select()
+            .from(BlobParts.TABLE_NAME)
+            .where(eq(BlobParts.ID, bindMarker(BlobParts.ID))));
+    }
+
+    private PreparedStatement prepareInsert(Session session) {
+        return session.prepare(insertInto(Blobs.TABLE_NAME)
+            .value(Blobs.ID, bindMarker(Blobs.ID))
+            .value(Blobs.POSITION, bindMarker(Blobs.POSITION))
+            .value(Blobs.PART, bindMarker(Blobs.PART)));
+    }
+
+    private PreparedStatement prepareInsertPart(Session session) {
+        return session.prepare(insertInto(BlobParts.TABLE_NAME)
+            .value(BlobParts.ID, bindMarker(BlobParts.ID))
+            .value(BlobParts.DATA, bindMarker(BlobParts.DATA)));
+    }
+
+    public CompletableFuture<Optional<BlobId>> save(byte[] data) {
+        if (data == null) {
+            return CompletableFuture.completedFuture(Optional.empty());
+        }
+        BlobId blobId = BlobId.forPayload(data);
+        return saveBlobParts(data, blobId)
+            .thenCompose(partIds -> saveBlobPartsReferences(blobId, partIds))
+            .thenApply(any -> Optional.of(blobId));
+    }
+
+    private CompletableFuture<Stream<Pair<Integer, PartId>>> saveBlobParts(byte[] data, BlobId blobId) {
+        return FluentFutureStream.of(
+            dataChunker.chunk(data, CHUNK_SIZE)
+                .map(pair -> writePart(pair.getRight(), blobId, pair.getKey())
+                    .thenApply(partId -> Pair.of(pair.getKey(), partId))))
+            .completableFuture();
+    }
+
+    private CompletableFuture<PartId> writePart(ByteBuffer data, BlobId blobId, int position) {
+        PartId partId = PartId.create(blobId, position);
+        return cassandraAsyncExecutor.executeVoid(
+            insertPart.bind()
+                .setString(BlobParts.ID, partId.getId())
+                .setBytes(BlobParts.DATA, data))
+            .thenApply(any -> partId);
+    }
+
+    private CompletableFuture<Stream<Void>> saveBlobPartsReferences(BlobId blobId, Stream<Pair<Integer, PartId>> stream) {
+        return FluentFutureStream.of(stream.map(pair ->
+            cassandraAsyncExecutor.executeVoid(insert.bind()
+                .setString(Blobs.ID, blobId.getId())
+                .setLong(Blobs.POSITION, pair.getKey())
+                .setString(Blobs.PART, pair.getValue().getId()))))
+            .completableFuture();
```

Comment:
When I think about that I think we store more information that needed.
Technicaly in the Blobs table we only need to store (Blobs.id, max(Position)) and then we can get all blobs part because position will be in [0, max(Position)].
Don't you think this optimization worst it?
Personally I only see benefits for it.